src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -291,50 +317,88 @@ void registerNext() {
+        sendControlChange(0, DEVICE_KNOB_STYLE + i, LED_STYLE_OFF);
+        ++i;
+      }
+    }
+
     void register(LXDeviceComponent device) {
-      if (this.device != device) {
-        unregister();
-        this.device = device;
-        if (this.device instanceof LXEffect) {
-          this.effect = (LXEffect) this.device;
-          this.effect.enabled.addListener(this);
-        } else if (this.device instanceof LXPattern) {
-          this.pattern = (LXPattern) this.device;
+      if (this.device == device) {
+        return;
+      }
+      unregister();
+      this.device = device;
+
+      if (this.device == null) {
+        clearKnobsAfter(0);
+        return;
+      } else if (this.device instanceof LXEffect) {
+        this.effect = (LXEffect) this.device;
+        this.effect.enabled.addListener(this);
+      } else if (this.device instanceof LXPattern) {
+        this.pattern = (LXPattern) this.device;
+      }
+
+      this.device.controlSurfaceSemaphore.increment();
+
+      boolean isEnabled = false;
+      if (this.effect != null) {
+        isEnabled = this.effect.isEnabled();
+      } else if (this.pattern != null) {
+        isEnabled = this.pattern == ((LXChannel) this.channel).getActivePattern();
+      }
+      sendNoteOn(0, DEVICE_ON_OFF, isEnabled ? LED_ON : LED_OFF);
+
+      int i = 0;
+      LXListenableParameter currentParentParameter = null;
+      for (LXListenableNormalizedParameter parameter : this.device.getRemoteControls()) {
+        if (parameter == null) {
+          this.knobs[i] = null;
+          sendControlChange(0, DEVICE_KNOB_STYLE + i, LED_STYLE_OFF);
+          continue;
         }
 
-        int i = 0;
-        boolean isEnabled = false;
-        if (this.device != null) {
-          if (this.effect != null) {
-            isEnabled = this.effect.isEnabled();
-          } else if (this.pattern != null) {
-            isEnabled = this.pattern == ((LXChannel) this.channel).getActivePattern();
-          }
-          for (LXListenableNormalizedParameter parameter : this.device.getRemoteControls()) {
-            if (i >= this.knobs.length) {
-              break;
-            }
-            this.knobs[i] = parameter;
-            if (parameter != null) {
-              parameter.addListener(this);
-              sendControlChange(0, DEVICE_KNOB_STYLE + i, parameter.getPolarity() == LXParameter.Polarity.BIPOLAR ? LED_STYLE_BIPOLAR : LED_STYLE_UNIPOLAR);
-              double normalized = (parameter instanceof CompoundParameter) ?
-                ((CompoundParameter) parameter).getBaseNormalized() :
-                parameter.getNormalized();
-              sendControlChange(0, DEVICE_KNOB + i, (int) (normalized * 127));
-            } else {
-              sendControlChange(0, DEVICE_KNOB_STYLE + i, LED_STYLE_OFF);
-            }
-            ++i;
+        if (i >= this.knobs.length) {
+          // Subtle: Even if all the knobs are full, we need to keep adding
```

Comment:
Just checking here... why do we need listeners for the children of the final knob? `AggregateParameter` should be a Listenable parameter, so listening to that ought to pick up the relevant changes that would make this knob have to change?

Wondering if perhaps the issue here is that I am not always firing a change event on LinkedColorParameter subparam changes, if the color isn't actually modified? Is that what this is playing defense against?

Reply:
Let's say we've filled up 7 knobs and the next parameter to be processed is the Hue of a new LCP. We assign the LCP to the 8th knob and set up a listener for the Hue parameter, so that when it changes later, we can update the knob's lights. But we need to do the same thing with saturation and brightness. Without this special case, on the next two loops when we're looking at the S and B params, we'd skip adding listeners because the 8 knobs are full.